SwiftSyntaxCLI/Sources/swift-syntax-cli/common/Builder.swift

```diff
@@ -0,0 +1,120 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+protocol Builder {
+  var arguments: BuilderArguments { get }
+}
+
+extension Builder {
+  func buildProduct(productName: String) throws {
+    print("** Building product " + productName + " **")
+    try build(packageDir: Paths.packageDir, productName: productName)
+  }
+
+  func buildExample(exampleName: String) throws {
+    print("** Building example " + exampleName + " **")
+    try build(packageDir: Paths.examplesDir, productName: exampleName)
+  }
+
+  private func build(packageDir: URL, productName: String) throws {
+    let args = ["--product", productName]
+
+    var env = ProcessInfo.processInfo.environment
+    env["SWIFT_BUILD_SCRIPT_ENVIRONMENT"] = "1"
+
+    if arguments.enableRawsyntaxValidation {
+      env["SWIFTSYNTAX_ENABLE_RAWSYNTAX_VALIDATION"] = "1"
+    }
+
+    if arguments.enableTestFuzzing {
+      env["SWIFTPARSER_ENABLE_ALTERNATE_TOKEN_INTROSPECTION"] = "1"
+    }
+
+    // Tell other projects in the unified build to use local dependencies
+    env["SWIFTCI_USE_LOCAL_DEPS"] = "1"
+    env["SWIFT_SYNTAX_PARSER_LIB_SEARCH_PATH"] = arguments.toolchain + "/lib/swift/macos"
+
+    try swiftpmInvocation(
+      action: "build",
+      packageDir: packageDir,
+      additionalArguments: args,
+      env: env
+    )
+  }
+
+  @discardableResult
+  func swiftpmInvocation(
+    action: String,
+    packageDir: URL,
+    additionalArguments: [String],
+    env: [String: String]
+  ) throws -> String {
+    var args = [action]
+    args += ["--package-path", packageDir.relativePath]
+    args += ["--scratch-path", arguments.buildDir]
+
+    #if !canImport(Darwin)
+    args += ["--enable-test-discovery"]
+    #endif
+
+    if arguments.release {
+      args += ["--configuration", "release"]
+    }
+
+    if let multirootDataFile = arguments.multirootDataFile {
+      args += ["--multiroot-data-file", multirootDataFile]
+    }
+
+    if arguments.disableSandbox {
+      args += ["--disable-sandbox"]
+    }
+
+    if arguments.verbose {
+      args += ["--verbose"]
+    }
+
+    args += additionalArguments
+
+    let process = Process()
+    let outputPipe = Pipe()
+    process.environment = env
+    process.executableURL = URL(fileURLWithPath: arguments.toolchain).appendingPathComponent("bin").appendingPathComponent("swift")
+    process.arguments = args
+    process.standardOutput = outputPipe
```

Comment:
I think it would be nice to wrap the `Process` API behind a custom abstraction layer, mostly because I don’t like how it works (for example how much overhead you need to write to get `stdout` and `stderr` or how you manually need to implement `PATH` lookup).

An API that I would like would be something like the following. I have not implemented the PATH lookup in it yet or tested it thoroughly but I think having something like this will make invoking other terminal commands a lot simpler everywhere.

```swift
/// Error thrown if a process terminates with a non-zero exit code.
struct NonZeroExitCodeError: Error {
  let stdout: String
  let stderr: String
  let exitCode: Int
}

/// Error thrown if `stdout` or `stderr` could not be decoded as UTF-8.
struct FailedToDecodeUTF8Error: Error {
  let data: Data
}

/// If `name` is not a fully qualified path, looks an executable with that name
/// up in `PATH` and runs it with `args`. If `additionalEnvironment` is specified,
/// the current environment will be augmented by these values.
///
/// Throws in either of these conditions:
///  - The executable could not be found or could not be run
///  - Stdout or stderr weren't valid UTF-8
///  - The process executes with a non-zero exit code.
@discardableResult
func runProcess(name: String, args: [String], additionalEnvironment: [String: String] = [:]) throws -> (stdout: String, stderr: String) {
  let process = Process()

  // Set up stdout
  let stdoutPipe = Pipe()
  let stdoutHandle = stdoutPipe.fileHandleForReading
  var stdoutData = Data()
  stdoutHandle.readabilityHandler = { pipe in
    stdoutData += pipe.availableData
  }
  process.standardOutput = stdoutPipe

  // Set up stderr
  let stderrPipe = Pipe()
  let stderrHandle = stderrPipe.fileHandleForReading
  var stderrData = Data()
  stderrHandle.readabilityHandler = { pipe in
    stderrData += pipe.availableData
  }
  process.standardError = stderrPipe

  // Set up environment
  var environment = ProcessInfo.processInfo.environment
  for (key, value) in additionalEnvironment {
    environment[key] = value
  }
  process.environment = environment

  // Set up arguments
  process.executableURL = /* do the path lookup dance I mentioned in a different comment */
  process.arguments = args

  // Run the process
  try process.run()
  process.waitUntilExit()

  guard let stdoutString = String(data: stdoutData, encoding: .utf8) else {
    throw FailedToDecodeUTF8Error(data: stdoutData)
  }
  guard let stderrString = String(data: stderrData, encoding: .utf8) else {
    throw FailedToDecodeUTF8Error(data: stderrData)
  }

  if process.terminationStatus != 0 {
    throw NonZeroExitCodeError(stdout: stdoutString, stderr: stderrString, exitCode: Int(process.terminationStatus))
  }

  return (stdoutString, stderrString)
}
```

Reply:
Could just grab ProcessRunner.swift from the stress tester. It also handles adding a wait for getting the rest of stdout/stderr.

Reply:
I picked the `ProcessRunner.swift` and trimmed it a bit so it just had what we needed.

Reply:
Great. It’s probably the better choice as well. Thanks @bnbarham for remembering me that it existed.